### PR TITLE
OSD-20596: osdctl  swarm secondary

### PR DESCRIPTION
**Pull request by amej**
osdctl command for outputting swarm info on Secondary

When Secondary triggers a swarm, they often manually create a list of current cards on the board which require work - ideally we can add this functionality to osdctl to remove that overhead.

Suggested command: osdctl swarm secondary

Reference of a quick implementation in bash: https://github.com/openshift/ops-sop/blob/master/v4/utils/secondary_open_tickets.sh

Acceptance Criteria:

    the command should output a list of In Progress OHSS tickets to be actioned by the on-call secondary

    the output is well formatted for Slack

What will you see here :eye_speech_bubble: 

Documentation:
Prerequisite:  
Env variable:  ` JIRA_API_TOKEN="SomethingGeneratedOnJiraSite"`

```
` dist/osdctl_linux_amd64_v1/osdctl -h`

CLI tool to provide OSD related utilities

Usage:
  osdctl [command]

Available Commands:
...
  swarm          Provides a set of commands for swarming activity
...

```

```
dist/osdctl_linux_amd64_v1/osdctl swarm -h                                                                                                                                           
                  
Provides a set of commands for swarming activity

Usage:
  osdctl swarm [command]

Available Commands:
  secondary   List unassigned JIRA issues based on criteria

Flags:
  -h, --help   help for swarm
Use "osdctl swarm [command] --help" for more information about a command.

```

```
dist/osdctl_linux_amd64_v1/osdctl swarm secondary -h                                                                                                                                                   

Lists unassigned Jira issues from the 'OHSS' project
		for the following Products
		- OpenShift Dedicated
		- Openshift Online Pro
		- OpenShift Online Starter
		- Red Hat OpenShift Service on AWS
		- HyperShift Preview
		- Empty 'Products' field in Jira
		with the 'Summary' field  of the new ticket not matching the following
		- Compliance Alert
		and the 'Work Type' is not one of the RFE or Change Request

Usage:
  osdctl swarm secondary [flags]

Examples:
#Collect tickets for secondary swarm
		osdctl swarm secondary

```
Actual command:
`dist/osdctl_linux_amd64_v1/osdctl swarm secondary`               
  gives output as

```

Timestamp:  2024-07-19 11:15:23.192569897 +0530 IST m=+6.621848548 Title: Swarm Secondary. 

>> OHSS Issues
[<Jira ID>|<jIRA url> ](Incident/High): <Summary>
- Created: 2024-07-18 11:27	Status: In Progress

```

### CODE REVIEW

File: cmd/swarm/secondary.go [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+	"github.com/spf13/cobra"
+)
+
+const (
+	DefaultProject = "OHSS"
+)
+
+var (
+	products = []string{"Openshift Dedicated", "Openshift Online Pro", "OpenShift Online Starter", "Red Hat OpenShift Service on AWS", "HyperShift Preview"}
+)
```
Comment (devppratik) [inline cmd/swarm/secondary.go:17]: var (
	products = []string{"\"Openshift Dedicated\"", "\"Openshift Online Pro\"", "\"OpenShift Online Starter\"", "\"Red Hat OpenShift Service on AWS\"", "\"HyperShift Preview\""}
)
```suggestion
var (
	products = []string{"Openshift Dedicated", "Openshift Online Pro", "OpenShift Online Starter", "Red Hat OpenShift Service on AWS", "HyperShift Preview"}
)
```

**Review by devppratik — CHANGES_REQUESTED** [review]

**Review by amej — COMMENTED** [review]
Thank you @devppratik for the comments.
In separate commit, I shall reformat the output as requested in the chat.

**Review by amej — COMMENTED** [review]
Changes accepted.

**Review by amej — COMMENTED** [review]
thank you.

File: cmd/swarm/secondary.go [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+func buildJQL() string {
+	productsJQL := fmt.Sprintf("Products in (%s)", strings.Join(products, ","))
+	// Build Summary and Work Type conditions
+	summaryCondition := `(summary !~ "Compliance Alert" AND status in (NEW) OR (summary ~ "Compliance Alert" AND status in (NEW)))`
+	workTypeCondition := `(("Work Type" != "Request for Change (RFE)" OR "Work Type" is EMPTY) AND status not in (Done, Resolved)) OR (status in (Done, Resolved) AND ("Work Type" != "Request for Change (RFE)" OR "Work Type" is EMPTY) AND resolutiondate > startOfDay(-2d))`
+
+	// Combine all conditions into final JQL
+	jql := fmt.Sprintf(`(project = %s AND %s AND (%s) AND (status in (New, "In Progress")) AND assignee is EMPTY)`, DefaultProject, productsJQL, summaryCondition+" AND "+workTypeCondition)
+
+	fmt.Printf("The query is ->  %s <- ", jql)
```
Comment (amej) [inline cmd/swarm/secondary.go:69]: @devppratik : The constructed query is:
`(project = OHSS AND Products in ("Openshift Dedicated","Openshift Online Pro","OpenShift Online Starter","Red Hat OpenShift Service on AWS","HyperShift Preview") AND ((summary !~ "Compliance Alert" AND status in (NEW) OR (summary ~ "Compliance Alert" AND status in (NEW))) AND (("Work Type" != "Request for Change (RFE)" OR "Work Type" is EMPTY) AND status not in (Done, Resolved)) OR (status in (Done, Resolved) AND ("Work Type" != "Request for Change (RFE)" OR "Work Type" is EMPTY) AND resolutiondate > startOfDay(-2d))) AND (status in (New, "In Progress")) AND assignee is EMPTY)`

File: cmd/swarm/secondary.go [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+	RunE: func(cmd *cobra.Command, args []string) error {
+
+		jiraClient, err := utils.GetJiraClient()
+		if err != nil {
+			return fmt.Errorf("failed to get Jira client: %w", err)
+		}
+		// Build JQL query
+		jql := buildJQL()
+		// Search jira issues
+		issues, _, err := jiraClient.Issue.Search(jql, nil)
```
Comment (amej) [inline cmd/swarm/secondary.go:45]: @devppratik : 
The constructed query within the jira ui does not give the following error.  Yet, this command gives the error

` Error: error fetching JIRA issues: The value 'OHSS' does not exist for the field 'project'.: request failed. Please analyze the request body for more details. Status code: 400`

**Review by amej — COMMENTED** [review]
@devppratik :  The code fails with `
 Error: error fetching JIRA issues: The value 'OHSS' does not exist for the field 'project'.: request failed. Please analyze the request body for more details. Status code: 400`

Please can you have a look ?

**Comment by devppratik** [comment]
Tested and Working

**Comment by devppratik** [comment]
/approve
Approving the PR because I have tested the code and all looks good

**Pull request merged**